rfd/00XY-access-own-sessions.md

```diff
@@ -0,0 +1,102 @@
+
+Properly supporting the new filter will require a migration procedure for the event table:
+
+1. Check if there is a `session.end` event stored without the `Participants` attribute, e.g. using a probe like:
+   ```
+   aws dynamodb scan \
+      --table-name Events \
+      --filter-expression "EventType = :name AND attribute_not_exists(Participants)" \
+      --expression-attribute-values '{":name":{"S":"session.end"}}'
+   ```
+1. If there is no such event, i.e. all stored `session.end` events already come with `Participants`, there is nothing to do.
+1. If such events exist, for each `session.end` event lacking the `Participants` attribute:
+   1. Get the list of participants from the event's `Fields` and store it as a string set under the `Participants` attribute.
+   1. Reupload the event similarly to [`migrateDateAttribute`](https://github.com/gravitational/teleport/blob/992c10f547a6b7c24247835d7711fadb46ad9022/lib/events/dynamoevents/dynamoevents.go#L1170).
+
+#### Firestore
+
+Our Firestore support filters event types by going through all the events (within a time range) and [unmarshaling the `Fields` on the Auth server side](https://github.com/gravitational/teleport/blob/992c10f547a6b7c24247835d7711fadb46ad9022/lib/events/firestoreevents/firestoreevents.go#L539-L550). Filtering by participants contained in the `session.end` fields can be easily performed in the same fashion.
+
+#### File log
+
+In order to filter by event type, the event file is processed and the individual entries unmarshaled at the Auth server. It should be straightforward to extend the procedure to support another field to filter by.
+
+### Remarks
+
+No new `teleport.yaml` options need to be introduced.
+
+Some Teleport deployments may depend on the legacy behavior of all users being able to join active sessions thanks to the implicit `KindSSHSession` privilege. However, since showing all active sessions to all users constitutes a potential security risk anyway, it should be preferable to get rid of the implicit privilege even in spite of the possibility of breaking an established use case.
+
+With this RFD it becomes impossible to make a user's own session recordings unavailable to the user as the RBAC checks get skipped in those scenarios.
```

Comment:
@klizhentas What do we want the UX to be here?

At the moment, customers configure Teleport to either see all sessions (for admins) or none (for users), do you want to maintain that behavior or are you okay to opt-in customers to now show own session by default?

If we don't want to introduce changes for existing customers, we'd have to support 3 modes:

1. You can see no sessions
2. You can only see your own sessions
3. You can see all sessions

That means if you are migrating an existing cluster, you would get either mode (1) or (3) with configuration allowing you to change to mode (2).

If you are bootstrapping a new cluster, you would get (2) by default with a choice to change to (1) or (3).

Reply:
It might be more appropriate to think in terms of only two modes:
A. RBAC checks aren't skipped when viewing own sessions (which is the current behaviour, i.e. you either see all or no sessions according to your privileges);
B. RBAC checks are skipped when viewing own sessions (i.e. you're able to see your own sessions even if you lack the privileges).

In the RFD's current formulation, since I wanted to avoid adding new `teleport.yaml` options and (B) seemed reasonable enough for all deployments, it's proposed that (B) is applied automatically as the new behaviour and without the ability to opt out.

If we wanted to allow an opt-out, the choice between (A) and (B) would probably have to be exposed as a `teleport.yaml` option.

Reply:
In connection with https://github.com/gravitational/teleport/pull/8035#discussion_r709514734 I came up with another approach to solve this.

The current proposal would allow a user to view his own sessions even if one of his roles explicitly denies the privilege (e.g. denying `list` for `session`). This is because the considered notion of "RBAC check failed" subsumes both "no allow rule match" and "a deny rule does match". It might make sense to distinguish these two cases and disable to access one's own sessions when the privilege is denied explicitly.

This could be a solution to
```
With this RFD it becomes impossible to make a user's own session recordings unavailable to the user as the RBAC checks get skipped in those scenarios.
```
which we so far discussed to resolve by adding a new configuration option. The allow/deny dichotomy may already make it possible to configure this scenario:
* If an allow rule matches, show all active sessions / recordings, as is already the case.
* If a deny rule matches, never show any active sessions / recordings.
* If none of the above (i.e. no applicable allow/deny rule is specified in any of the user's roles), show only the sessions the user is involved in.